src/backend/distributed/connection/connection_configuration.c

```diff
@@ -0,0 +1,335 @@
+ *
+ * This function must be called before others in this file, though calling it
+ * after use of the initialized ConnParams structure will result in any
+ * populated parameter settings being lost.
+ */
+void
+InitConnParams()
+{
+	Size maxSize = CalculateMaxSize();
+	ConnParamsInfo connParams = {
+		.keywords = malloc(maxSize * sizeof(char *)),
+		.values = malloc(maxSize * sizeof(char *)),
+		.size = 0,
+		.maxSize = maxSize
+	};
+
+	memset(connParams.keywords, 0, maxSize * sizeof(char *));
+	memset(connParams.values, 0, maxSize * sizeof(char *));
+
+	ConnParams = connParams;
+}
+
+
+/*
+ * ResetConnParams frees all strings in the keywords and parameters arrays,
+ * sets their elements to null, and resets the ConnParamsSize to zero before
+ * adding back any hardcoded global connection settings (at present, only the
+ * fallback_application_name of 'citus').
+ */
+void
+ResetConnParams()
+{
+	Index paramIdx = 0;
+
+	for (paramIdx = 0; paramIdx < ConnParams.size; paramIdx++)
+	{
+		free((void *) ConnParams.keywords[paramIdx]);
+		free((void *) ConnParams.values[paramIdx]);
+
+		ConnParams.keywords[paramIdx] = ConnParams.values[paramIdx] = NULL;
+	}
+
+	ConnParams.size = 0;
+
+	AddConnParam("fallback_application_name", "citus");
+}
+
+
+/*
+ * AddConnParam adds a parameter setting to the global libpq settings according
+ * to the provided keyword and value. Under assert-enabled builds, array bounds
+ * checking is performed.
+ */
+void
+AddConnParam(const char *keyword, const char *value)
+{
+	Assert((ConnParams.size + 1) < ConnParams.maxSize);
+
+	ConnParams.keywords[ConnParams.size] = strdup(keyword);
+	ConnParams.values[ConnParams.size] = strdup(value);
+	ConnParams.size++;
+
+	ConnParams.keywords[ConnParams.size] = ConnParams.values[ConnParams.size] = NULL;
+}
+
+
+/*
+ * CheckConninfo is a building block to help implement check constraints and
+ * other check hooks against libpq-like conninfo strings. In particular, the
+ * provided conninfo must:
+ *
+ *   - Not use a uri-prefix such as postgres:// (it must be only keys and values)
+ *   - Parse using PQconninfoParse
+ *   - Only set keywords contained in the provided whitelist
+ *
+ * This function returns true if all of the above are satisfied, otherwise it
+ * returns false. If the provided errmsg pointer is not NULL, it will be set
+ * to an appropriate message if the check fails.
+ *
+ * The provided whitelist must be sorted in a manner usable by bsearch, though
+ * this is only validated during assert-enabled builds.
+ */
+bool
+CheckConninfo(const char *conninfo, const char **whitelist,
+			  Size whitelistLength, char **errorMsg)
+{
+	PQconninfoOption *optionArray = NULL;
+	PQconninfoOption *option = NULL;
+	Index whitelistIdx PG_USED_FOR_ASSERTS_ONLY = 0;
+
+	/*
+	 * If the user doesn't need a message, just overwrite errmsg with a stack
+	 * variable so we can always safely write to it.
+	 */
+	if (errorMsg == NULL)
+	{
+		char *errorMsgString;
```

Comment:
This seems wrong. Once this block is over this variable is implicitly popped off the stack. There are other blocks below that declare variables that will overwrite the memory of this one. Maybe move the variable to the top of the function.

Reply:
Wow talk about a complete slip-up. I was thinking "stack == function call" but you're totally right that automatic storage rules mean this is scoped to this bracketed region. Fixing it.

Reply:
Fixed.